### ISSUE TITLE
InodeTable in reverse mode can lead to IV changes and reuse of IVs

### ISSUE THREAD

**Issue by slackner** [rfjakob/gocryptfs#398]
This issue already exists for quite some time, but I don't think there is a bug report for it yet. I was initially hoping to fix it together with #158, but since we haven't come up with a good solution for that one yet, I'm opening this as a separate bug report.

In reverse mode, all inodes with `Nlink > 1` are tracked in a map `inodeTable`. The map is never cleared, and not preserved across restarts of `gocryptfs`. This logic can lead to several unexpected side-effects:

* Assume the source directory contains two files referencing the same inode, then the encrypted file content depends on which path is accessed first after `gocryptfs` is started. Depending on the backup software this could be completely random (e.g., if multiple threads are used to enumerate the directory content), and could cause frequent retransmits. rsync is also affected when `--checksum` is used.

* Assume a file is deleted, and some time later (without restarting gocryptfs) a new file is created, which has the same inode number. Since `gocryptfs` never clears the `inodeTable`, the new file reuses the initialization vectors of the previous file, even when it is located at a completely different location.

**Comment by rfjakob** [comment]
For (1), I wonder if it would be better to drop the IV (or use all zeros) for files that have hard links. As reverse mode uses AES-SIV, having unique IVs is not critical.

**Comment by slackner** [comment]
I don't think that would be better. In fact, it would make it even easier for an attacker to mess with a backup, when all hardlinked files use the same IV.

For (1) my proposed solution would be to derive the IV directly from the inode number, which is the same, no matter which path is used. However, that would not really solve the second problem. And it would mean that creating a hardlink at some later time also causes a full retransmit of the file.

**Comment by slackner** [comment]
What also would be an option: Drop the whole `inodeTable` logic, and instead generate unique inode numbers. Then the content is transferred multiple times, but at least it is fully deterministic, and does not depend on the access order of files/directories.

**Comment by rfjakob** [comment]
> derive the IV directly from the inode number

The problem here is when you move your data to a new filesystem (user gets a bigger hdd for example), all the encrypted content changes.

**Comment by digitalsignalperson** [comment]
For (2), under what condition does this lead to a full compromise? Knowledge of the plaintext of two files encrypted with the same IV?

That might not be too hard given non-obfuscated file sizes, and inode reuse happens pretty quick. E.g. I've seen deleting a file on XFS, it can be the next file or 2nd next new file re-using the inode number.

Here's a possible scenario I can imagine:
- Maintainer of a code repo syncs all their files, including project folders to some "untrusted" offsite storage, relying on gocryptfs deterministic mode
- an attacker also has access to the encrypted files
- no file size obfuscation so the attacker can easily identify files. e.g. we know they maintain project X, know the file structure, can easily identify specific project and build files
- the build process creates a few hardlinks. Attacker identifies these and saves a copy
- clean build deletes the hardlink files
- IV is reused relatively soon as new files are created. Attacker can monitor the files and when it sees the clean build, check every new file with hopes they can both identify it based on the project structure, and that it was the file that on the maintainer's host that recycled the inode.
- now we have copies of two files we know the plaintext and used the same IV

Thoughts?

Example poking around repos I have cloned: lanzaboot project creates 3 pairs of hard links to various build-script-build, build-script-main, etc during build.

**Comment by rfjakob** [comment]
Yes it's as you said in the edit.

In reverse mode, IV reuse happens all the time - not only for hardlinked
files. That's why reverse mode only uses AES-SIV, which is resistant
against IV reuse, but unfortunately much slower than AES-GCM.
